### DISCUSSION TITLE
Can't get Browsershot to work on Windows with XAMPP, few different errors

### DISCUSSION THREAD

**Discussion by gengenacc** [spatie/browsershot#728] [Q&A]
Hello

I am trying to make Spatie Browsershot work on Windows with XAMPP.
Windows 11
node v18.16.0
XAMPP - latest, with PHP 8.2.4
puppeteer - v20.2.0
spatie/browsershot": v3.57

First of all I installed `puppeteer`:

```
npm install -g puppeteer
```

At first, I wanted to output the result directly to the browser, so I had the following code in the controller:
```
use Spatie\Browsershot\Browsershot;

    public function createPDFWithBrowsershot(Request $request)
    {
        Browsershot::html('<h1>Test</h1>')->pdf();
    }
```

But then I get the following error:
> The command "node ^"C:^\xampp^\htdocs^\my_project^\vendor^\spatie^\browsershot^\src/../bin/browser.js^ ..."

So then I tried to do the following (though not sure it's a good option for Windows, but it did make the above error disappear):
```
    public function createPDFWithBrowsershot(Request $request)
    {
        Browsershot::html('<h1>Test</h1>')->setNodeBinary('PATH %~dp0;%PATH%;')->pdf();
    }
```
And the error above disappeared, however the page was blank, so something seems to still not work, but no errors.
Edit: When I `dd(Browsershot::html('<h1>Test</h1>')->setNodeBinary('PATH %~dp0;%PATH%;')->pdf());`, I get `""` (empty string)..

So instead of trying to output the PDF to the browser, I tried to save it as a pdf file:
```
    public function createPDFWithBrowsershot(Request $request)
    {
        $save_to_file = 'C:/pdfs/file.pdf';
        Browsershot::html('<h1>Test</h1>')->setNodeBinary('PATH %~dp0;%PATH%;')->savePdf($save_to_file);
    }
```
But I got a new error:
>For some reason Chrome did not write a file at `C:/pdfs/file.pdf`. Command ======= [] Output ======

I also tried several other paths:
```
 $save_to_file = 'C:\pdfs\file.pdf';
```
or
```
$save_to_file = storage_path('pdfs/file.pdf');
```
But I get the same error.

I even tried uninstalling the global puppeteer and installing it only in my project:
```
npm install puppeteer
```

And even installing chromium:
```
npm install chromium
```

Nothing makes it work

Why?

**Comment by kabouter** [comment]
Did you manage to fix this?

**Comment by lance2k** [comment]
My friend is having the same problem with xampp/browsershot, did you manage to fix this?

**Comment by RyanGSampson** [comment]


**Reply by RyanGSampson** [reply]
I solved it by pushing to Forge on a testing server and running it on the web.

**Comment by kabouter** [comment]
I solved this by using Laravel Sail on Windows with wsl2.

**Reply by mahmoud-757** [reply]
Hello, I am using sail and wsl2, how did you manage to solve it?
Any help is appreciated.

**Reply by kabouter** [reply]
your project is accessible after you sail up?(through localhost or webadres?)

If you want to use composer or npm with sail you must precede the keyword "composer" or "npm" with the word sail in your CLI.

for example:
`npm install puppeteer `
should be
`sail npm install puppeteer`

It is bit too long ago to give a full walkthrough.

**Comment by zaidpirwani** [comment]
Hello,
So I have been trying to generate PDFs and was trying the new laravel-pdf package which uses browsershot.

Was getting error
Error Output: ================ 'node' is not recognized as an internal or external command, operable program or batch file.

which was fixed by adding path to node, but still had issues
```
return Pdf::view('pdf.purchaseOrder', [
'record' => $record,
'record_quote' => $record->quotes->where('is_selected', true)->first(),
])->format('A4')
->withBrowsershot(function (Browsershot $browsershot) {
$browsershot->setNodeBinary("C:/nodejs/node.exe");
})
->save($record->getPRNumber() . '.pdf');
```

Next error was quite vague and toook me some time to work it out
`The command ""C:/nodejs/node.exe" "C:\Users\Zaid\PhpstormProjects\LaravelHello\vendor\spatie\browsershot\src/../bin/browser.cjs" "{""url"":""https://example.com"",""action"":""screenshot"",""options"":{""type"":""png"",""path"":""a.png"",""args"":[],""viewport"":{""width"":800,""height"":600},""newHeadless"":true}}"" failed. Exit Code: 1(General error) Working directory: C:\Users\Zaid\PhpstormProjects\LaravelHello\public Output: ================ Error Output: ================ [Error: ENOENT: no such file or directory, mkdtemp 'undefined\temp\puppeteer_dev_chrome_profile-XXXXXX'] { errno: -4058, code: 'ENOENT', syscall: 'mkdtemp', path: 'undefined\temp\puppeteer_dev_chrome_profile-XXXXXX' }
`

**### System / Package Versions**
node, v20.9.0
npm, v10.1.0
php, v8.2.4
laravel, v10.43.0
laravel-pdf, v1.1.2
Windows 11 Pro, version 23H2
PhpStorm 2023.3.3, Build #PS-233.14015.96, built on January 24, 2024
my composer.lock file has browsershot at ^4.0, "spatie/browsershot": "^4.0",


node available in windows path, accessible from cmd prompt / terminal

Was getting error as mentioned here [laravel-pdf Bug#70](https://github.com/spatie/laravel-pdf/issues/70)

I have commented by journey of tracing the issue potentially to browsershot not sending env to process that runs puppeteer, not sure it is due to browsershot or how artisan serve works - I understand this problem is not occuring on Linux and other services, but I really needed to run on windows as this is my dev platform - though I deploy on ubuntu machines mostly.

after much code diving, it seemed that the process of puppeteer being run by browsershot was not getting any environment variables, I traced the process creation to line 1032 in file vendor/spatie/browsershot/src/Browsershot.php
`$process = $this->isWindows() ? new Process($fullCommand) : Process::fromShellCommandline($fullCommand);
`
and modified it to send env using getenv() method
`$process = $this->isWindows() ? new Process($fullCommand,null,getenv() ) : Process::fromShellCommandline($fullCommand);
`
now I am able to generate PDF Files. I know am not supposed to modify files in vendor folders - need to check further

my pdf code is 
```
Pdf::view('pdf.purchaseOrder', [
    'record' => $record,
    'record_quote' => $record->quotes->where('is_selected', true)->first(),
])
    ->format('A4')
    ->margins(10,10,10,10)
    ->withBrowsershot(function (Browsershot $browsershot) {
        $browsershot
            ->setNodeBinary("C:/nodejs/node.exe")
            ->setOption('newHeadless', true);
    })
    ->save($record->getPRNumber() . '.pdf');
```

also am sending the whole env of laravel, not sure if that is right or not - am currently running this locally

**Reply by zaidpirwani** [reply]
sharing here, hoping it helps someone, I wanted to start a new discussion, but this seems similar enough to add here - apologies for reviving an old thread.

am interested in learning how would this be further traced and fixed ?

is this a browsershot issue or more upstream ?

**Reply by christmex** [reply]
Awesome man

**Reply by ThantHtetAung101** [reply]
You're the savior, man T - T

**Reply by SyedMuradAliShah** [reply]
Thanks savior, I'm try to limit the ENV

```
 $envVars = [
            "LOCALAPPDATA"  => getenv("LOCALAPPDATA"),
            "Path"                     => getenv("Path"),
            "SystemRoot"         => getenv("SystemRoot"),
            "USERPROFILE"      => getenv("USERPROFILE"),
        ];
        
$process = $this->isWindows() ? new Process($fullCommand, null, $envVars) : Process::fromShellCommandline($fullCommand);
```

**Comment by ghost** [comment]
Still not working in Laravel 10
